[Rules] Improve processing rules speed

TD-er opened the pull request:
On ESP32 we do have enough free memory for keeping all rules lines in memory.
On ESP8266 we can save some time by keeping the file handle open.

Still to do:
- [x] Test on ESP8266
- [x] Add index cache for mapping events to positions in the file(s)
- [x] Optimize parsing rules lines
- [x] Check if we can stop parsing rules when an event was handled
- [x] Try to cache frequent happing events, even on ESP8266
- [x] Document special case of events with leading `!` and `*` in the `on ... do` line of rules.
- [x] Add `%eventname%` and `%eventpar%` to access the event itself in rules.
- [x] Allow for more than 4 `%eventvalueN%` 
- [x] Add ``%eventvalueX|Y%`` with: X = event value nr > 0, Y = default value when eventvalue does not exist. N.B. default value can be a string, thus ``%eventvalue3|[int#3]%`` should be possible as long as the default value not contains neither ``|`` nor ``%``.
- [x] Allow for empty eventvalues, like ``myEvent=1,,3,4``
- [x] Allow for strings as eventvalues, which makes complex event calls possible like: ``myEvent=[int#1]``.

On ESP32 the rules parsing is improved quite a bit, but still it is relatively slow.
But at least the always given advice to add an `on ... do ... endon` block at the start of the first rules file now actually does speed up rules parsing.
Not sure if that ever was a speed-up or sometimes got broken, or perhaps (most likely) was planned to be implemented and then forgotten....

Main changes in this PR:
- Removed the limit of max. 4 eventvalues.
- Match using wildcards.  Such a wildcard match is also needed when processing string eventvalues (at least the first eventvalue)
- When no eventvalues are present in the event, these will be replaced by `0` (later will make an option to store what should be the replacement text)

Newly added variables to be used in rules: 
- `%eventname%` = The event upto the `=` character, thus the entire event if no `=` is found.
- `%eventpar%` = The last part of `%eventname%` after the first `#` character. Thus empty if no `#` is present.


Resolves: #1812
Resolves: #2285
Resolves: #2787 
Resolves: #3511
Resolves: #3741
Resolves: #3791
Resolves: #3999

TD-er committed:
[Rules cache] Add handler for speeding up rules parsing

TD-er committed:
[Rules] Let the RulesHelper parse line by line

TD-er committed:
[Rules] Cache rules per line on ESP32

TD-er committed:
[Rules] Only parse lines that need to be parsed

Also stop parsing rules when event is handled.

TD-er committed:
[Rules] Nesting level does not matter to stop parsing rules file

tonhuisman commented:
> by keeping the file handle open

Won't that cause issues, as there are only 5 file-handles available? (Max open files: 5, according to the Info page) But I assume that can be increased.

TD-er commented:
> > by keeping the file handle open
> 
> Won't that cause issues, as there are only 5 file-handles available? (Max open files: 5, according to the Info page) But I assume that can be increased.

That's why we need to test ;)
But you're right, that can (and should) be set to a higher value.
And I think on ESP8266 opening a file isn't taking that long as those builds have small SPIFFS file systems (larger file system slows down things considerably)
On 8M LittleFS opening a file takes 20 - 30 msec, but since we only use that on ESP32, that delay is mitigated by having the rules in memory.
However, it still needs quite a lot of parsing time and I got the feeling this parsing is slower on the ESP32 compared to the ESP8266 even though the ESP32 does run at 3x the clock speed.

ghtester commented:
I am also afraid of possible decreased stability... Does not mean that more open file handles is more RAM consumed?
From my perspective the current rule speed processing is great.

TD-er commented:
> I am also afraid of possible decreased stability... Does not mean that more open file handles is more RAM consumed? From my perspective the current rule speed processing is great.

If the rules parsing is fast enough, we can disable the open file issues.

TD-er committed:
[WiFi] Rate limit checking WiFi.isConnected()

TD-er committed:
[Rules] Small tweaks to speed up processing rules

TD-er committed:
[Rules] Rate limit calls to background tasks

Sometimes the background tasks were ran on reading every line of rules code.
However running it multiple times per msec is way overkill.

TD-er committed:
[Rules] Minor tweaks to improve rules execution speed

TD-er committed:
[Rules] Cache which rules lines start with "on " for ESP32

Reduces seek time to find the next rules block with rules loaded in memory (thus only ESP32)

TD-er committed:
[Rules] Fix missing include

TD-er committed:
[Rules] Add caching to speed up rules processing

TD-er committed:
[Rules] Fix build issue on ESP32

TD-er committed:
[Rules] Fix build on ESP32

TD-er committed:
[Rules] Fix missing include on ESP32

TD-er committed:
[Rules] Improve speed processing rules + option to disable rules cache

TD-er committed:
[PlatformIO] Update ESP32 PIO platform to 3.5.0

3.3.x no longer possible to build on Windows after full clean of PlatformIO

TD-er commented:
Changes made today:
- Add index cache to immediately jump to the block in the rules matching the event
- Add option to disable this cache (default enabled) if for whatever reason this might not work well in some situations. (e.g. matching events based on system variables)
- On ESP8266 only keep the file handle open for the last used rules file. Thus having only one rules file may save time opening file.
- Dynamic sort rules event cache based on how often it was matched, so most frequent matched event will be tested first.
- Tweak rules parsing to improve speed

All-in-all, the processing speed for rules has been improved considerably.
Depending on the number of rules blocks in your rules files and overall load of the system, the number of processed events per second can be as high as 100 events per sec.


Would be highly appreciated if this can be tested by a number of people to see if something got broken.

tonhuisman commented:
I've been testing this on a copy of a production system (my door/chime controller) but with actual actions and events disabled, to avoid disturbing the family 😛 but some of the events won't fire, like this one: `on Doorcontact#State do` while a similarly complex event: `on Bellbutton#State do` works as expected.
After changing the taskname and  event handler to read `on Deurcontact#State do` it works as intended, ~I suspect it has to do with [these lines of code](https://github.com/letscontrolit/ESPEasy/pull/4004/files#diff-2a2730a623effea016b8004255e1745e205b5836c9df2b92ab4c2c68b13caa8dR28-R31)~ Haven't found a good reason for this issue yet, even after a good hour of searching and testing. Not sure what's blocking this 'event names starting with do' issue yet

TD-er commented:
Only thing I can think of is that it might have some special characters in the rules line?
At boot, or when you save something, the rules files are being read again and some info level logs are given with the position in what file each on....do line is found.

Do you have duplicates of these lines? Or maybe there are multiple blocks which can match this?

TD-er commented:
Ah now I understand what's wrong there...
I need to start searching for " do" starting at the position after the found "on ", not from the start of the line.
And maybe even better, I need to have some string inbetween first or else "on  Doorbel...." (2 spaces between 'on' and 'Doorbel') may act strange.

TD-er committed:
[Rules] fix parsing events starting with " do"

TD-er commented:
Should be fixed and also made some minor tweaks to slightly improve execution speed

tonhuisman commented:
Tested and confirmed that it is now fixed. Only when using 3 or more spaces between `on` and `Doorcontact` it will fail, but that is to be expected based on the code, and would be odd anyway.

TD-er commented:
Will address that issue too, by checking the event inbetween is a valid string.

N.B. I'm now also implementing the `%eventname%` (everything before the `=` sign) and using wildcards in the rules events.

TD-er committed:
[Rules] Add %eventname% and %eventpar% to allow for simpler rules

TD-er committed:
Rate limit the number of background task calls to improve speed

Calling it upto 200x a second is more than enough.

TD-er committed:
[Timing Stats] Add process_system_event_queue() to the timing stats

TD-er committed:
[Rules] Add setting to disable cached event reorder

It is unknown whether reordering of events in the cache may have negative side effects, so better to have an option to disable it for special cases.

TD-er committed:
[Rules] Allow for more than 4 %eventvalueN%

TD-er committed:
[Rules] Document using > 4 eventvalues + using string type eventvalues

TD-er commented:
This PR now also removes the limit of max. 4 eventvalues.
And you can now also match using wildcards.
Such a wildcard match is also needed when processing string eventvalues (at least the first eventvalue)
Newly added variables to be used in rules: 
- `%eventname%` = The event upto the `=` character, thus the entire event if no `=` is found.
- `%eventpar%` = The last part of `%eventname%` after the first `#` character. Thus empty if no `#` is present.

When no eventvalues are present in the event, these will be replaced by `0` (later will make an option to store what should be the replacement text)

TD-er committed:
[Rules] Add %eventvalue0% to get all arguments at once.

TD-er committed:
[Rules] Fix compiler warning compare signed/unsigned

TD-er committed:
[Rules] Check for invalid %eventvalueN% occurences

TD-er committed:
[Rules] Add documentation on %eventvalue0%

TD-er committed:
[Rules] Allow for empty event values

Example rules:

```
on eventvalues* do
  logentry,"test eventvalues: 0:%eventvalue% 1:%eventvalue1% 2:%eventvalue2% 3:%eventvalue3% 4:%eventvalue4% 5:%eventvalue5% 6:%eventvalue6%"
  logentry,"All eventvalues: %eventvalue0%"
endon
```

Test event: `event,eventvalues=1, '',     ,4444 ,5,6`

Logs:

```
523694 : Info   : EVENT: eventvalues=1, '',     ,4444 ,5,6
523703 : Info   : ACT  : logentry,"test eventvalues: 0:1 1:1 2: 3: 4:4444 5:5 6:6"
523706 : Info   : test eventvalues: 0:1 1:1 2: 3: 4:4444 5:5 6:6
523708 : Info   : ACT  : logentry,"All eventvalues: 1, '',     ,4444 ,5,6"
523710 : Info   : All eventvalues: 1, '',     ,4444 ,5,6
```

TD-er committed:
[Rules] Add option for default values on non-existing eventvalues

tonhuisman commented on docs/source/Rules/Rules.rst:
```diff
@@ -326,15 +374,117 @@ It can also be used in rules. Every occurance of this text will then be replaced
+``%eventvalueN%`` - substitutes the N-th event value (everything that comes after
+the '=' sign).
+
+``%eventvalue%`` will be replaced by 
```
This line seems not finished yet?

TD-er replied:
I guess so ;)
Maybe someone just messaged me just when I was typing the documentation with some more good feedback on this issue ;)

TD-er committed:
[Rules] Add 'restrict' keyword to safely allow executing %eventvalueN%

TD-er committed:
[Rules] Fix issue when processing eventvalue without default value

AloisKlingler commented:
I would like to try this on my ESP32 testdevice for shutter control, it has in total nearly 20kBytes rules, I think I would test a lot changes from here with this.
where can I fetch a build for testing? I would need normal_ESP32_4M316k :-) <-- found the build in the meanwhile :-)
Thanks :-)

TD-er commented:
@AloisKlingler Would be great to hear from your tests.
Maybe you can also fetch some timing stats from before and after these changes on the same system running for at least an hour or so (depending on how frequent you get events)

TD-er commented:
Just for the record, the last Github actions build is here: https://github.com/letscontrolit/ESPEasy/actions/runs/2200906952
You need to be logged in to be able to download the zip file.

TD-er committed:
[Rules] Update documentation regarding event values.

AloisKlingler commented:
this build is for me like easter and christmas at the same time 👍 🥇
Timing stats as screenshots attached for comparison between 20220328 and 20220421.
in the screenshot statistics for firing events and their processing time. calculated. Expected: finishsz 28 seconds, all other 18 seconds. the "overall stop" is timer-driven and should be in this case also 28 seconds.
20220421's values are really frightening precise.
20220328' values are like I mentioned in https://www.letscontrolit.com/forum/viewtopic.php?p=57837#p57837 not useable.
the currently used 20210114 version has values about +20%  (so e.g. 35 seconds instead of the 28 seconds and 22 seconds instead of the 18 seconds), but there it affects only the rule parsing speed noticeably, not as in 20220328 where also the pulsed GPIO timers are disturbed by about 100% (e.g. 4 seconds pulsed instead of 2 seconds).

for reference also the 4 rulesets are attached. the "Let,70,..." to "Let,82,..." are in those rulesets if it if for interest. I fired in both cases **event,openall** and measured multiple times (I did NOT measure the first run!) with 20220328 I got quite different results, but all of them were always far away from the aimed 18 / 28 seconds. with 20220421 the time difference was on all runs nearly the same.

I think I will plug in the ESP with the new build tomorrow so I will be able to tell you also maybe negative effects. so far it was stable and I did not notice any negative.

Thanks. :-)

rules:
[rules.zip](https://github.com/letscontrolit/ESPEasy/files/8532070/rules.zip)

release 20220421:
![20220421](https://user-images.githubusercontent.com/11095760/164485509-cc2cebad-3ab4-4f3a-b13a-1c421ed600ba.png)

release 20220328:
![20220328](https://user-images.githubusercontent.com/11095760/164485515-a014ad58-67bd-421f-a248-684801ef4b5c.png)

TD-er commented:
Just browsed through your rules.
That's some extensive set of rules.
Nice to see the parser still manages to process them rather quickly.
And I think your rulesset may become extremely more readable when we have "named variables", so `[var#1]` can be something like `[var#uptime]` (not sure about the syntax yet, just thinking about it, will for sure not be part of this PR)
